text/0000-maybe-dangling.md

```diff
@@ -0,0 +1,247 @@
+// start dropping memory it points to).
+fn thread(control: ..., closure: MaybeDangling<impl FnOnce() + 'lifetime>) {
+    closure.into_inner()();
+    control.signal_done();
+    // A lot of time can pass here.
+}
+```
+
+The 3rd example can be fixed by storing the `buffer` inside a `MaybeDangling`, which disables its aliasing requirements:
+
+```rust=
+struct OwningRef<U, T: StableDeref<Target=U>> {
+    buffer: MaybeDangling<T>,
+    ref_: NonNull<U>, // conceptually borrows from `buffer`.
+}
+```
+
+As long as the `buffer` field is not used, the pointer stored in `ref_` will remain valid.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The standard library contains a type `MaybeDangling<P>` that is safely convertible with `P` (i.e., the safety invariant is the same), and that has all the same niches as `P`, but that does allow passing around dangling boxes and references within unsafe code.
+
+"Behavior considered undefined" is adjusted as follows:
+
+```diff
+- * Breaking the pointer aliasing rules. `&mut T` and `&T` follow LLVM’s
+-   scoped noalias model, except if the &T contains an UnsafeCell<U>.
++ * Breaking the pointer aliasing rules. `Box<T>`, `&mut T` and `&T` follow LLVM’s
++   scoped noalias model, except for `UnsafeCell<_>` inside the `T`. 
++   References must not be dangling while they are live, again except for
++   `UnsafeCell<_>` inside the `T`. (The exact liveness duration is not
++   specified, but it is certainly upper-bounded by the syntactic lifetime
++   assigned by the borrow checker. When a reference is passed to a function,
++   it is live at least as long as that function call.) All this also
++   applies when values of these types are passed in a field of a compund
++   type, except behind pointer indirections and when the pointers or
++   references are inside `MaybeDangling`.
+[...]
+ * Producing an invalid value, even in private fields and locals. 
+   "Producing" a value happens any time a value is assigned to or 
+   read from a place, passed to a function/primitive operation or 
+   returned from a function/primitive operation. The following 
+   values are invalid (at their respective type):
+[...]
+-  * A reference or Box<T> that is dangling, unaligned, or points to an
+-    invalid value.
++  * A reference or `Box<T>` that is unaligned or null, or whose pointee
++    type `T` is uninhabited. Furthermore, except when this value occurs
++    inside a `MaybeDangling`, if the reference/`Box<T>` is dangling or points
++    to an invalid value, it is itself invalid.
+```
+
+*Note: this might seem to alter the aliasing rules compared to the current reference more than just by adding a `MaybeDangling` exception (specifically when it talks about the liveness duration of references), but really it just clarifies semnatics we have applied since Rust 1.0, and incorporates [#98017](https://github.com/rust-lang/rust/pull/98017).*
```

Comment:
Why do we need to incorporate #98017? It is unrelated to this RFC, no? I think it is better to keep things separated.

Reply:
It would be strange to write incorrect documentation here by omitting that clause. Really the problem is that the reference is rather outdated and so the diff has to cover much more than just what this RFC changes.

I'll see if I can get the reference updates, then the diff here would shrink.

Reply:
To make things less confusing, the diff is now relative to https://github.com/rust-lang/reference/pull/1290.